### PR TITLE
Daemon Hooking

### PR DESCRIPTION
**Pull request by Camerooooon**

### CODE REVIEW

**Comment by JakeRoggenbuck** [comment]
Yay! very well done!

**Comment by Camerooooon** [comment]
The solution that I decided to use was to replace most of the references to the Daemon with references to an Arc<Mutex<Daemon>> which is what I would say is a "wrapper struct" that "surrounds" the daemon. This struct controls which thread is allowed to read and write to the object (in this case the daemon). A thread can request to read/write data by performing a "lock" on the data which basically says to whatever thread is requesting the data that it needs to wait for the main thread to stop using the data (in this case if the main thread is utilizing the data the socket thread will wait for the request to finish before proceeding thereby holding up the socket thread which is not a problem since our acs client can be patient).

**Comment by Camerooooon** [comment]
I've implemented the command listening and parsing. You can now use socat to broadcast `0|test hello` following the protocol outlines.

**Comment by Camerooooon** [comment]
Ok, this PR is *almost* complete. It needs a LOT of reviews there are so many situations where it can break and it's quite unstable.

**Comment by Shuzhengz** [comment]
its still a draft pr

File: src/daemon.rs
Comment (Shuzhengz) [inline src/daemon.rs:420]: just a little cursed lol

File: src/network.rs
Comment (Shuzhengz) [inline src/network.rs:36]: i miss javadocs

**Review by Shuzhengz — COMMENTED** [review]
also don't forget about cargo fmt

File: src/daemon.rs
Comment (Camerooooon) [reply src/daemon.rs:420]: I honestly don't see a better way to do it though! Do you have any suggestions?

Comment (JakeRoggenbuck) [reply src/daemon.rs:420]: Yea, that is fair, it probably is the best solution

Comment (JakeRoggenbuck) [reply src/daemon.rs:420]: Here are some good resources I have yet to read, but will soon

- https://www.rustnote.com/concurrency/mutex.html
- https://itsallaboutthebit.com/arc-mutex/

I'm not that great at concurrency in rust yet but I'll do some reading, this seems way less cursed than much of how concurrency happens in languages other than rust, so that's good

File: src/network.rs
Comment (JakeRoggenbuck) [inline src/network.rs:106]: Yay tests!

**Comment by Camerooooon** [comment]
I'm going to make acs monit broadcast and listen for the hello packet before merge.

**Comment by Camerooooon** [comment]
We can add more packets in later prs. Also I need to add more error handling.

**Comment by JakeRoggenbuck** [comment]
Sounds good, great work!

**Comment by JakeRoggenbuck** [comment]
So basically, remove the run method from daemon, replace it with a run function that accepts a safe mutable version of the daemon. That's cool

File: src/daemon.rs
Comment (JakeRoggenbuck) [inline src/daemon.rs:550]: I am a little worried about the speed of this

Comment (JakeRoggenbuck) [reply src/daemon.rs:550]: I guess we have a sleep right after it, maybe it does not matter here

File: src/network.rs
Comment (JakeRoggenbuck) [inline src/network.rs:68]: Wow, this is mega smart, now I see why threads are important here. Zero blocking, smart

File: src/daemon.rs
Comment (Camerooooon) [reply src/daemon.rs:558]: It doesn't matter much, the daemon is only "locked" for an incredibly small amount of time while data is being read.

File: src/network.rs
Comment (Camerooooon) [reply src/network.rs:73]: Yup it made stuff complicated but overall it should be efficient.

File: src/network.rs
Comment (JakeRoggenbuck) [inline src/network.rs:75]: I think this for loop should just pass stream to a function, like deal_with_stream(stream);

**Comment by Camerooooon** [comment]
> This looks very close to being done, if not already done. Anything else you think would be good to add?

It needs to be stabilized lots of unhandled exceptions.

**Comment by Camerooooon** [comment]
Almost done, just fixed a major bug!

File: src/network.rs
Comment (Camerooooon) [reply src/network.rs:75]: Oki!

**Comment by Camerooooon** [comment]
Looks like we are go for review @JakeRoggenbuck and @Shuzhengz! The daemon hooking is working, please re-review then merge this!

**Comment by JakeRoggenbuck** [comment]
> Looks like we are go for review @JakeRoggenbuck and @Shuzhengz! The daemon hooking is working, please re-review then merge this!

Awesome!

**Comment by JakeRoggenbuck** [comment]
What is a good way to test the sending to the socket?

**Comment by Camerooooon** [comment]
Cargo monit attempts to send the hello packet. It will then print out the response to std out ( you will need to scroll up because it gets pushed away by the GUI ) you can test by reading that response. Additionally you can use socat to manually attach to the sock and send packets yourself.

**Comment by JakeRoggenbuck** [comment]
Tested, everything seems good.

**Review by JakeRoggenbuck — APPROVED** [review]
Great!

**Pull request merged**
